[cmHuang777] iP

cmHuang777 opened the pull request:
## DukePro
>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind from having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
- download it from [here](https://github.com/cmHuang777/ip).
- double-click it.
- add your tasks.
- let it manage your tasks for you 👍 

Ans it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Duke bot = new Duke("data/tasks.txt");
        bot.run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

cmHuang777 committed:
Modified greetings and exit

cmHuang777 committed:
added function to echo command

cmHuang777 committed:
Added function to store text and list them

cmHuang777 committed:
Added functions mark tasks as done or undone

cmHuang777 committed:
added toDos,, Events, Deadlines to the type of tasks

cmHuang777 committed:
Added Automated Text UI Testing

cmHuang777 committed:
Added Error Handling

cmHuang777 committed:
Added function to delete tasks

cmHuang777 committed:
Add function to save tasks to disk

Tasks list will be stored in a file locally. Any modification to the list of tasks will also update the file.

cmHuang777 committed:
Merge branch 'branch-Level-7'

cmHuang777 committed:
Change the date format to java LocalDate

Previously the date format is in String, users can just type anything.
There was no input checking/ or was difficult. Changing the date format
makes the date more standardised.

The input String for event /from, /to and deadline /by,
now only take in the format of "yyyy-mm-dd".
Other format will be considered as invalid input.
The valid string will be parsed as java LocalDateObject.

cmHuang777 committed:
Merge branch 'branch-Level-8'

cmHuang777 committed:
Add more classes to support OOP

Previously all logic is done in the Duke class. More OOP style was needed.

This commit adds the following classes:
- Command
- DukeException
- Parser
- Storage
- TaskList
- Ui

The respective logic is moved to the respective classes, to support OOP.
The Duke class is now a class that handles the main logic of the program.

However, Command could still be broken down into more classes, and there
may also be other classes that could be added to better support OOP.

cmHuang777 committed:
Merge branch 'branch-A-MoreOOP'

cmHuang777 committed:
put all source files in duke package

cmHuang777 committed:
Merge branch 'branch-A-Packages'

cmHuang777 committed:
Merge branch 'add-gradle-support'

cmHuang777 committed:
Add gradle support

cmHuang777 committed:
Merge branch 'branch-A-Gradle'

cmHuang777 committed:
Add some JUnit tests

cmHuang777 committed:
Merge branch 'branch-A-JUnit'

cmHuang777 committed:
Enable jar file creation

cmHuang777 committed:
Add function to find tasks based on description

cmHuang777 committed:
Add javadoc to all main classes

cmHuang777 committed:
Merge branch 'branch-Level-9'

cmHuang777 committed:
Clean up code style

cmHuang777 committed:
Merge branch-A-CodingStandard into master

cmHuang777 committed:
Merge branch-A-javadoc into master

tayruxin commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,71 @@
+public abstract class Task {
+
+    protected final String task;
+    protected boolean done;
```
Should the naming of this boolean be renamed to isDone instead?

tayruxin commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public class Todo extends Task {
+
+    private static final String TYPE = "[T]";
```
Good job for naming this constant correctly by having it as all caps!

tayruxin commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,79 @@
+ * Object to deal with user interaction.
+ */
+public class Ui {
+    private static final String CHAT_BOT_NAME = "Genos";
```
Good job for using all upper case and underscore to separate the words

tayruxin commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,229 @@
+     */
+    public boolean execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        Task task = null;
+        switch (this.action) {
+        case INVALID:
```
Good job on adhering to the correct indexing for the switch statement

tayruxin commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,67 @@
+package duke;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
```
Good job for not using wildcard imports

tayruxin reviewed (commented):
Overall, the code follows the coding standard well. The naming of variables and spacing are done well. Overall, a very good job.

cmHuang777 committed:
Update code to match style check

cmHuang777 committed:
Merge branch 'branch-A-checkStyle'

cmHuang777 committed:
Add GUI to the project.

Done using the JavaFX tutorial from SE-EDU.

Current CLI interface will no longer work, please use the GUI instead.

cmHuang777 committed:
Add acknowledgement in code

cmHuang777 committed:
Merge branch 'branch-Level-10'

cmHuang777 committed:
Add assertions to the code

Purpose is to ensure that no unexpected behaviour occurs.

Assertions are used to check that the state of the program is as expected.
Mostly checking for null values, in case some methods are called
with null values.

cmHuang777 committed:
Improve overall code CodeQuality

It is done to address some comments from the PR review.
I have also spotted some poor code quality and fixed them.

Some changes include:
* Fixing some code smells
* Making variable names follow the convention
* Making the code more readable
* Simplifying complex boolean expressions

cmHuang777 committed:
TaskList: Modify to use Streams in implementation

As per the requirment in streams, the TaskList class has been modified
to use streams in its implementation. Using streams also gives shorter
code and better readability.

Most of the for loops were replaced with IntStream in the TaskList class.
It is probably the easiest way to convert a for loop to a stream.

No real impact on performance was observed. The chatbot should still be
able function as per normal.

cmHuang777 committed:
Merge pull request #1 from cmHuang777/branch-A-Assertions

Add assertions to the code

cmHuang777 committed:
Update branch with master

cmHuang777 committed:
Merge pull request #2 from cmHuang777/branch-A-CodeQuality

Improve overall code CodeQuality

cmHuang777 committed:
Merge branch 'master' into branch-A-Streams

cmHuang777 committed:
Merge pull request #3 from cmHuang777/branch-A-Streams

Modify to use Streams in implementation

cmHuang777 committed:
Use GitHub Actions to set up Continuous Integration

cmHuang777 committed:
Merge pull request #4 from cmHuang777/branch-A-CI

Use GitHub Actions to set up Continuous Integration

cmHuang777 committed:
Add extra location feature

As per requirement, the user can now add a location to a task.

The user can add a location to a task by supplementing the command
with /at <location>. User can also use location <place> command to find
all tasks at that <place>. /at is optional and will not affect previous
functionality.

Other changes: Refactor code to follow OOP principles. Separated
command class into many different commands classes. Also moved the
classes into respective packages.
There should not have any impact on previous functionality.

cmHuang777 committed:
Merge pull request #5 from cmHuang777/branch-D-Places

Add extra location feature

cmHuang777 committed:
Improve some code quality

cmHuang777 committed:
Add user guide and fix minor bugs

cmHuang777 committed:
Fix style errors

cmHuang777 committed:
Rename png in docs folder

cmHuang777 committed:
Fix minor GUI bug
